src/Compiler/Checking/CheckExpressions.fs

```diff
@@ -10874,17 +10874,23 @@ and TcNormalizedBinding declKind (cenv: cenv) env tpenv overallTy safeThisValOpt
                 errorR(Error(FSComp.SR.tcLiteralCannotHaveGenericParameters(), mBinding))
             
         if g.langVersion.SupportsFeature(LanguageFeature.EnforceAttributeTargetsOnFunctions) && memberFlagsOpt.IsNone && not attrs.IsEmpty then
-            let rhsIsFunction = isFunTy g overallPatTy
-            let lhsIsFunction = isFunTy g overallExprTy
-            let attrTgt =
-                match rhsIsFunction, lhsIsFunction with
-                | false, false when declaredTypars.IsEmpty -> AttributeTargets.Field ||| AttributeTargets.Property ||| AttributeTargets.ReturnValue
-                | _, _ -> AttributeTargets.Method ||| AttributeTargets.ReturnValue
-
-            TcAttributesWithPossibleTargets false cenv env attrTgt attrs |> ignore
+            TcAttributeTargetsOnLetBindings cenv env attrs overallPatTy overallExprTy (not declaredTypars.IsEmpty)
 
         CheckedBindingInfo(inlineFlag, valAttribs, xmlDoc, tcPatPhase2, explicitTyparInfo, nameToPrelimValSchemeMap, rhsExprChecked, argAndRetAttribs, overallPatTy, mBinding, debugPoint, isCompGen, literalValue, isFixed), tpenv
 
+// Note:
+// - Let bound values can only have attributes that uses AttributeTargets.Field ||| AttributeTargets.Property ||| AttributeTargets.ReturnValue
+// - Let function bindings can only have attributes that uses AttributeTargets.Method ||| AttributeTargets.ReturnValue
+and TcAttributeTargetsOnLetBindings (cenv: cenv) env attrs overallPatTy overallExprTy areTyparsDeclared =
+    let rhsIsFunction = isFunTy cenv.g overallPatTy
+    let lhsIsFunction = isFunTy cenv.g overallExprTy
+    let attrTgt =
+        match rhsIsFunction, lhsIsFunction with
+        | false, false when not areTyparsDeclared -> AttributeTargets.Field ||| AttributeTargets.Property ||| AttributeTargets.ReturnValue
+        | _, _ -> AttributeTargets.Method ||| AttributeTargets.ReturnValue
```

Comment:
I see that all you did was move this code, but even so, the logic might be a little easier to understand like this:

```suggestion
    let attrTgt =
        if
            // It's a type function:
            //     let x<'a> = …
            areTyparsDeclared
            // It's a regular function-valued binding:
            //     let f x = …
            //     let f = fun x -> …
            || isFunTy cenv.g overallPatTy
            || isFunTy cenv.g overallExprTy
        then
            AttributeTargets.ReturnValue ||| AttributeTargets.Method
        else
            AttributeTargets.ReturnValue ||| AttributeTargets.Field ||| AttributeTargets.Property
```

Reply:
Yeah much clear. Thanks for the review